mono/mini/mini-posix.c

```diff
@@ -1106,22 +1106,22 @@ mono_dump_native_crash_info (const char *signal, MonoContext *mctx, MONO_SIG_HAN
 	dump_memory_around_ip (mctx);
 }
 
+#ifdef HAVE_UCONTEXT_H
+#include <ucontext.h>
+#endif
+
 void
 mono_post_native_crash_handler (const char *signal, MonoContext *mctx, MONO_SIG_HANDLER_INFO_TYPE *info, gboolean crash_chaining, void *context)
 {
-	if (!crash_chaining) {
-		/*Android abort is a fluke, it doesn't abort, it triggers another segv. */
-#if defined (HOST_ANDROID)
-		exit (-1);
-#else
-		abort ();
-#endif
-	}
-	mono_chain_signal (info->si_signo, info, context);
+	if (crash_chaining)
+		mono_chain_signal (info->si_signo, info, context);
 
-	// we remove Mono's signal handlers from crashing signals in mono_handle_native_crash(), so re-raising will now allow the OS to handle the crash
-	// TODO: perhaps we can always use this to abort, instead of explicit exit()/abort() as we do above
+	// we remove Mono's signal handlers from crashing signals in mono_handle_native_crash(), so restoring the context or re-raising will now allow the OS to handle the crash
+#ifdef HAVE_UCONTEXT_H
+	setcontext((ucontext_t *) context);
```

Comment:
> If the context was obtained by a call to a signal handler, then old standard text says that "program execution continues with the program instruction following the instruction interrupted by the signal". However, this sentence was removed in SUSv2, and the present verdict is "the result is unspecified".

https://man7.org/linux/man-pages/man2/setcontext.2.html

That doesn't give a whole lot of confidence that this will work.  The "old standard" behavior makes it sound like this is going to resume the program after the bad instruction, meaning you're not going to get the signal again.  The "present verdict" is ¯\\_(ツ)_/¯

Also if it does work, will it work (ie restore the stack) when we're using `sigaltstack`?

Reply:
Yeah I wanted to give this a shot, it should be clear from crash tests on CI if this could actually work (and where)

The altstack case should probably be handled here: https://github.com/mono/mono/blob/master/mono/mini/mini-runtime.c#L3571

Reply:
Some trivia:

- C99 (7.14.3) claims that returning from a signal handler for `SIGFPE`, `SIGILL`, and `SIGSEGV` is UB. POSIX doesn't say much about this, or about the meaning or valid range of uses of the ucontext_t argument to an SA_SIGINFO handler.
- `setcontext` is apparently prohibited on watchOS and tvOS (https://github.com/apple/darwin-libplatform/blob/master/include/ucontext.h#L39).
- The libplatform signal trampoline on watchOS doesn't look like it'll pass a `ucontext_t` at all to the signal handler (https://github.com/apple/darwin-libplatform/blob/master/src/setjmp/generic/sigtramp.c#L96).
- The actual behavior of XNU seems to be that the `ucontext_t` passed in as the third parameter to an SA_SIGINFO handler will be used as the context that is restored by `sigreturn` (which is automatically called by the libplatform signal trampoline), the data pointed to by the third parameter is mutable, and the kernel will restore whatever IP we set in the context. All of this machinery should correctly handle alternate signal stacks (https://github.com/opensource-apple/xnu/blob/master/bsd/dev/i386/unix_signal.c#L663), so we ought to be able to just return normally from the signal handler after uninstalling mono-specific signal handlers.

Reply:
The way we handle nullrefs etc, is we set up some stuff on the stack, then change the ip in the sigcontext to point to some mono code and return from the signal handler.   This doesn't work on watchos etc.